common/preconditioner/isai_kernels.hpp.inc

```diff
@@ -125,22 +129,26 @@ __forceinline__ __device__ void generic_generate(
                 [&](IndexType, IndexType m_idx, IndexType i_idx,
                     config::lane_mask_type, bool valid) {
                     if (valid) {
-                        trisystem(nz, i_idx) = m_values[m_row_begin + m_idx];
+                        if (m_col_idxs[m_row_begin + m_idx] < row && col == row)
+                            atomic_add(&rhs_one_idxs[row % subwarps_per_block],
+                                       size_type{1});
```

Comment:
I would suggest you [(sub)warp-aggregate](https://developer.nvidia.com/blog/cuda-pro-tip-optimized-filtering-warp-aggregated-atomics/) the results, then you should be able to avoid using atomics and shared memory:
```cpp
count += popcnt(subwarp.ballot());
```
Alternative: Store partial counts and reduce afterwards
```cpp
local_count += (m_col_idxs[m_row_begin + m_idx] < row && col == row);
// ...
count = reduce(subwarp, local_count)
```